File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/dungeon/Dungeon.java
package com.joeshuff.dddungeongenerator.generator.dungeon;

import android.graphics.Point;
import com.joeshuff.dddungeongenerator.GeneratingActivity;
import com.joeshuff.dddungeongenerator.util.Logs;
import com.joeshuff.dddungeongenerator.generator.features.RoomFeature;
import com.joeshuff.dddungeongenerator.generator.features.StairsFeature;
import com.joeshuff.dddungeongenerator.generator.floors.DungeonSection;
import com.joeshuff.dddungeongenerator.generator.floors.Floor;
import com.joeshuff.dddungeongenerator.generator.generating.MinSpanningTree;
import com.joeshuff.dddungeongenerator.generator.monsters.Bestiary;

import java.util.*;

public class Dungeon {

	transient public static int MAP_SIZE = 800;

	public String name;

	public int startX;
	public int startY;
	public int endX;
	public int endY;
	public int width;
	public int height;

	Environment.ENVIRONMENT_TYPE selectedEnvironment;

	Creator.CREATOR dungeonCreator;

	Purpose.PURPOSE dungeonPurpose;

	History.HISTORY dungeonHistory;

	Modifier globalModifier = new Modifier();

	Modifier userModifier = new Modifier();

	transient Random rnd; //SEED THIS IF YOU WANT THE SAME RESULTS
	String seed = "";

	List<Floor> dungeonFloors = new ArrayList<>();

	//Generation all complete
	transient boolean allCompleted = false;

//	HashMap<Integer, Giffer> floorGifs = new HashMap<>();
//	Giffer fainGiffer = new Giffer(0);

	transient GeneratingActivity activity;

	public Dungeon() {}

	public Dungeon (GeneratingActivity c, int startX, int startY, int endX, int endY) {
		activity = c;

		this.startX = startX;
		this.startY = startY;
		this.endX = endX;
		this.endY = endY;
		
		this.width = endX - startX;
		this.height = endY - startY;

		Bestiary.launchBestiary(c);
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getDungeonDescription() {
        String description = "You find yourself in a dungeon " + getSelectedEnvironment().getDescription().toLowerCase() + ".\n";

        if (getDungeonCreator() == Creator.CREATOR.NO_CREATOR) {
            description = description + "You discover that this dungeon is naturally formed.";
        } else {
            description = description + "This dungeon seems to be created by " + getDungeonCreator().getDescription() + "\n";
            description = description + "They created this dungeon as a " + getDungeonPurpose().getTitle().toLowerCase() + "\n" + getDungeonPurpose().getDescription();
        }

        return description;
    }

    public void setUserModifier(Modifier modifier) {
		this.userModifier = modifier;
	}

	public void setSeed(String seed) {
		String numberseed = "";

		for (char c : seed.toCharArray()) {
			int cVal = c;
			numberseed = numberseed + cVal;
		}

		this.seed = seed;
        rnd = new Random(stringToSeed(numberseed));
		Logs.i("Dungeon", "seed is : " + seed, null);
	}

	public static long stringToSeed(String s) {
		if (s == null) {
			return 0;
		}
		long hash = 0;
		for (char c : s.toCharArray()) {
			hash = 31L*hash + c;
		}
		return hash;
	}

	public void setRoomSize(int percentage) {
		Room.changeRoomSize(percentage / 100f);
	}

	public void setLongCorridors(boolean longCorridors) {
		MinSpanningTree.setLongCorridors(longCorridors);
	}

	public void setLinearProgression(boolean linearProgression) {
		DungeonSection.linearProgression = linearProgression;
	}

	public Environment.ENVIRONMENT_TYPE getSelectedEnvironment() {
		return selectedEnvironment;
	}

	public Creator.CREATOR getDungeonCreator() {
		return dungeonCreator;
	}

	public Purpose.PURPOSE getDungeonPurpose() {
		return dungeonPurpose;
	}

	public History.HISTORY getDungeonHistory() {
		return dungeonHistory;
	}

	public String getSeed() {
		return seed;
	}

	public void generateAttributes() {
		selectedEnvironment = Environment.generateEnvironmentType(rnd);
		dungeonCreator = Creator.generateCreator(rnd);

		List<Modifier> modifiers = new ArrayList<>();
		modifiers.add(userModifier);
		modifiers.addAll(Arrays.asList(selectedEnvironment.getModifier(), dungeonCreator.getModifier()));

		if (dungeonCreator != Creator.CREATOR.NO_CREATOR) {
			dungeonPurpose = Purpose.getPurpose(rnd);
			modifiers.add(dungeonPurpose.getModifier());

			dungeonHistory = History.getHistory(rnd);
			modifiers.add(dungeonHistory.getModifier());
		}

		globalModifier = Modifier.combineModifiers(modifiers, true);

		name = NameGenerator.generateName(this);
	}

	public void generate() {
		generateAttributes();

		Floor firstFloor = new Floor(this, rnd, 0);
		firstFloor.fillFloor();
//        firstFloor.splitFloor();

		dungeonFloors.add(firstFloor);

		branchOut();

		calculateNearestPartner();

		calculateRejectedRooms();

		triangulate();

		minSpanningTree();

		combinePaths();

		clearUnnecessaryData();

		pathFind();

		finalise();

		complete();
	}

    public Floor addFloorForLevel(int level) {
        Optional<Floor> theFloor = dungeonFloors.stream().filter(floor -> floor.getLevel() == level).findFirst();

        if (theFloor.isPresent()) {
            return theFloor.get();
        }

        Floor newFloor = new Floor(this, rnd, level);
        dungeonFloors.add(newFloor);

        return newFloor;
    }

    public Floor getLowestDungeonFloor() {
		Floor lowest = getDungeonFloors().get(0);

		for (Floor floor : getDungeonFloors()) {
			if (floor.getLevel() < lowest.getLevel()) {
				lowest = floor;
			}
		}

		return lowest;
	}

	public Floor getHighestDungeonFloor() {
		Floor highest = getDungeonFloors().get(0);

		for (Floor floor : getDungeonFloors()) {
			if (floor.getLevel() > highest.getLevel()) {
				highest = floor;
			}
		}

		return highest;
	}

	public Floor getDungeonFloorAtLevel(int level) {
		for (Floor f : getDungeonFloors()) {
			if (f.getLevel() == level) return f;
		}

		return null;
	}

	private void branchOut() {
        activity.setProgressText("Generating Rooms...");

		for (int i = 0; i < dungeonFloors.size(); i ++) {
			Logs.i("Dungeon", "generating rooms for floor " + i, null);
			dungeonFloors.get(i).branchOut();
		}
	}

    private void calculateNearestPartner() {
        activity.setProgressText("Spreading out Rooms...");

	    for (Floor floor : dungeonFloors) {
            for (DungeonSection section : floor.getSectionList()) {
                section.calculateNearestPartner();
            }
        }
    }

    private void calculateRejectedRooms() {
        for (Floor floor : dungeonFloors) {
            for (DungeonSection section : floor.getSectionList()) {
                section.calculateRejectedRooms();
            }
        }
    }

    private void triangulate() {
		activity.setProgressText("Triangulating Rooms...");

        for (Floor floor : dungeonFloors) {
            for (DungeonSection section : floor.getSectionList()) {
                section.triangulate();
            }
        }
    }

    private void minSpanningTree() {
		activity.setProgressText("Finding best corridors...");

        for (Floor floor : dungeonFloors) {
            for (DungeonSection section : floor.getSectionList()) {
                section.minSpanningTree();
            }
        }
    }

    private void combinePaths() {
        for (Floor floor : dungeonFloors) {
            for (DungeonSection section : floor.getSectionList()) {
                section.combinePaths();
            }
        }
    }

    private void pathFind() {
		activity.setProgressText("Connecting Rooms...");

        for (Floor floor : dungeonFloors) {
            for (DungeonSection section : floor.getSectionList()) {
                section.pathFind();
            }
        }
    }

    private void finalise() {
		activity.setProgressText("Generation Complete");

        for (Floor floor : dungeonFloors) {
            for (DungeonSection section : floor.getSectionList()) {
                section.finalise();
            }

			int i = 1;
			for (Room r : floor.getAllRooms()) {
				r.setId(i);
				i ++;
			}
        }
    }

    private void clearUnnecessaryData() {
        for (Floor floor : dungeonFloors) {
            for (DungeonSection section : floor.getSectionList()) {
                section.clearUnnecessaryData();
            }
        }
    }

	public void complete() {
	    allCompleted = true;

	    List<StairsFeature> stairsToDecide = new ArrayList<>();

		for (Floor f : dungeonFloors) {
		    for (Room r : f.getAllRooms()) {
		        for (RoomFeature rf : r.getFeatureList()) {
		            if (rf instanceof StairsFeature) {
                        stairsToDecide.add((StairsFeature) rf);
                    }
                }
            }
        }

        for (StairsFeature sf : stairsToDecide) sf.getConnectedRoom();

        activity.runOnUiThread(() ->activity.onCompleted());
	}

	public Room getRoomAt(int floor, int x, int y) {
	    for (Room r : getDungeonFloors().get(floor).getAllRooms()) {
	        if (r.containsGlobalPoint(new Point(x, y))) return r;
        }

        return null;
    }

	public Random getRnd() {
		return rnd;
	}

	public Modifier getGlobalModifier() {
		return globalModifier;
	}

    public List<Floor> getDungeonFloors() {
        return dungeonFloors;
    }
}

File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/dungeon/Dungeon.kt
package com.joeshuff.dddungeongenerator.generator.dungeon

import com.joeshuff.dddungeongenerator.db.models.Point
import com.joeshuff.dddungeongenerator.generator.dungeon.Environment.ENVIRONMENT_TYPE
import com.joeshuff.dddungeongenerator.generator.dungeon.History.HISTORY
import com.joeshuff.dddungeongenerator.generator.dungeon.Purpose.PURPOSE
import com.joeshuff.dddungeongenerator.generator.features.StairsFeature
import com.joeshuff.dddungeongenerator.generator.floors.DungeonSection
import com.joeshuff.dddungeongenerator.generator.floors.Floor
import com.joeshuff.dddungeongenerator.generator.generating.MinSpanningTree
import com.joeshuff.dddungeongenerator.memory.MemoryGeneration
import com.joeshuff.dddungeongenerator.screens.home.DungeonHistoryItem
import com.joeshuff.dddungeongenerator.util.Logs
import io.reactivex.ObservableEmitter
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.Ignore
import io.realm.annotations.PrimaryKey
import java.text.SimpleDateFormat
import java.util.*

open class Dungeon(
        @PrimaryKey var id: Int
): RealmObject() {
    companion object {
        @Transient var MAP_SIZE = 800

        fun stringToSeed(s: String?): Long {
            if (s == null) {
                return 0
            }
            var hash: Long = 0
            for (c in s.toCharArray()) {
                hash = 31L * hash + c.toLong()
            }
            return hash
        }
    }

    private var name: String = ""

    private var startX = 0
    private var startY = 0
    private var endX = 0
    private var endY = 0
    var width = 0
    var height = 0

    private var environment: String? = null
    var selectedEnvironment: ENVIRONMENT_TYPE?
        get() { return environment?.let { ENVIRONMENT_TYPE.valueOf(it) }?: null }
        set(new) { new?.let { environment = it.name } }

    var dungeonCreator: Creator? = null
        private set

    private var purpose: String? = null
    var dungeonPurpose: PURPOSE?
        get() { return purpose?.let { PURPOSE.valueOf(it) }?: null }
        set(new) { new?.let { purpose = it.name } }

    private var history: String? = null
    var dungeonHistory: HISTORY?
        get() { return history?.let { HISTORY.valueOf(it) }?: null }
        set(new) { new?.let { history = it.name } }


    private var globalModifier = Modifier()
    private var userModifier = Modifier()

    @Ignore
    lateinit var rnd: Random

    private var seed = ""

    private var dungeonFloors: RealmList<Floor> = RealmList()

    //Generation all complete
    var allCompleted = false

    constructor(): this((System.currentTimeMillis() / 1000).toInt(), 0, 0, 0, 0, "")

    constructor(startX: Int, startY: Int, endX: Int, endY: Int, seed: String): this((System.currentTimeMillis() / 1000).toInt(), startX, startY, endX, endY, seed)

    constructor(id: Int, startX: Int, startY: Int, endX: Int, endY: Int, seed: String): this(id) {
        setSeed(seed)

        this.startX = startX
        this.startY = startY
        this.endX = endX
        this.endY = endY

        width = endX - startX
        height = endY - startY
    }

    fun getName() = name

    fun getDungeonDescription(): String {
        var description = ""

        selectedEnvironment?.let {
            description += "You find yourself in a dungeon ${it.description.toLowerCase()} .\n"
        }

        dungeonCreator?.let {creator ->
            if (creator.creatorType == Creator.CREATOR.NO_CREATOR) {
                description += "You discover that this dungeon is naturally formed."
            } else {
                dungeonPurpose?.let { purpose ->
                    description += "This dungeon seems to be created by ${creator.getDescription()} \n"
                    description += "They created this dungeon as a ${purpose.title.toLowerCase()}\n ${purpose.description} \n"
                }
            }
        }

        return description
    }

    fun setUserModifier(modifier: Modifier) {
        userModifier = modifier
    }

    fun setSeed(seed: String?) {
        var numberseed = ""
        for (c in (seed?: "").toCharArray()) {
            val cVal = c.toInt()
            numberseed += cVal
        }
        this.seed = seed?: ""
        rnd = Random(stringToSeed(numberseed))
        Logs.i("Dungeon", "seed is : ${this.seed}", null)
    }

    fun setRoomSize(percentage: Int) {
        Room.changeRoomSize(percentage / 100f)
    }

    fun setLongCorridors(longCorridors: Boolean) {
        MinSpanningTree.setLongCorridors(longCorridors)
    }

    fun setLinearProgression(linearProgression: Boolean) {
        DungeonSection.linearProgression = linearProgression
    }

    fun getSeed(): String {
        return seed
    }

    override fun equals(other: Any?): Boolean {
        return if (other !is Dungeon) false else id == other.id
    }

    fun generateAttributes() {
        rnd?.let {randomness ->
            selectedEnvironment = Environment.generateEnvironmentType(randomness)
            dungeonCreator = Creator.generateCreator(randomness)

            val modifiers: ArrayList<Modifier> = ArrayList()
            modifiers.add(userModifier)

            selectedEnvironment?.modifier?.let { modifiers.add(it) }
            dungeonCreator?.creatorType?.modifier?.let { modifiers.add(it) }

            if (dungeonCreator?.creatorType != Creator.CREATOR.NO_CREATOR) {
                dungeonPurpose = Purpose.getPurpose(randomness)
                dungeonPurpose?.modifier?.let { modifiers.add(it) }
                dungeonHistory = History.getHistory(randomness)
                dungeonHistory?.modifier?.let { modifiers.add(it) }
            }

            globalModifier = Modifier.combineModifiers(modifiers, true)
        }

        name = NameGenerator.generateName(this)
    }

    fun generate(publishSubject: ObservableEmitter<String>) {
        generateAttributes()

        val firstFloor = Floor(this, rnd, 0)
        firstFloor.fillFloor()

        dungeonFloors.add(firstFloor)

        branchOut(publishSubject)

        calculateNearestPartner(publishSubject)

        calculateRejectedRooms()

        triangulate(publishSubject)

        minSpanningTree(publishSubject)

        combinePaths()

        clearUnnecessaryData()

        pathFind(publishSubject)

        finalise(publishSubject)

        complete()

        publishSubject.onComplete()
    }

    fun addFloorForLevel(level: Int): Floor {
        val theFloor = dungeonFloors.firstOrNull { floor: Floor -> floor.level == level }
        if (theFloor != null) return theFloor

        val newFloor = Floor(this, rnd, level)
        dungeonFloors.add(newFloor)
        return newFloor
    }

    val lowestDungeonFloor: Floor?
        get() {
            return getDungeonFloors().minBy { it.level }
        }

    val highestDungeonFloor: Floor?
        get() {
            return getDungeonFloors().maxBy { it.level }
        }

    fun getDungeonFloorAtLevel(level: Int): Floor? {
        return getDungeonFloors().firstOrNull { it.level == level }
    }

    private fun branchOut(publishSubject: ObservableEmitter<String>) {
        publishSubject.onNext("Generating Rooms...")
        var i: Int = 0

        while (i < getDungeonFloors().size) {
            val thisFloor = getDungeonFloors()[i]
            Logs.i("Dungeon", "generating rooms for floor ${thisFloor.level}", null)
            thisFloor.branchOut()
            i ++
        }
    }

    private fun calculateNearestPartner(publishSubject: ObservableEmitter<String>) {
        publishSubject.onNext("Spreading out Rooms...")
        getDungeonFloors().forEach { floor ->
            floor.sectionList.forEach { section ->
                section.calculateNearestPartner()
            }
        }
    }

    private fun calculateRejectedRooms() {
        getDungeonFloors().forEach { floor ->
            floor.sectionList.forEach { section ->
                section.calculateRejectedRooms()
            }
        }
    }

    private fun triangulate(publishSubject: ObservableEmitter<String>) {
        publishSubject.onNext("Triangulating Rooms...")
        getDungeonFloors().forEach { floor ->
            floor.sectionList.forEach { section ->
                section.triangulate()
            }
        }
    }

    private fun minSpanningTree(publishSubject: ObservableEmitter<String>) {
        publishSubject.onNext("Finding best corridors...")
        getDungeonFloors().forEach { floor ->
            floor.sectionList.forEach { section ->
                section.minSpanningTree()
            }
        }
    }

    private fun combinePaths() {
        getDungeonFloors().forEach { floor ->
            floor.sectionList.forEach { section ->
                section.combinePaths()
            }
        }
    }

    private fun pathFind(publishSubject: ObservableEmitter<String>) {
        publishSubject.onNext("Connecting Rooms...")
        getDungeonFloors().forEach { floor ->
            floor.sectionList.forEach { section ->
                section.pathFind()
            }
        }
    }

    private fun finalise(publishSubject: ObservableEmitter<String>) {
        publishSubject.onNext("Generation Complete")
        getDungeonFloors().forEach { floor ->
            floor.sectionList.forEach { section ->
                section.finalise()
            }

            //Renumber all the rooms so that they are actually sequentially numbered
            var i = 1
            for (r in floor.allRooms) {
                r.id = i
                i++
            }
        }
    }

    private fun clearUnnecessaryData() {
        getDungeonFloors().forEach { floor ->
            floor.sectionList.forEach { section ->
                section.clearUnnecessaryData()
            }
        }
    }

    fun complete() {
        allCompleted = true
        val stairsToDecide: ArrayList<StairsFeature> = ArrayList()
        getDungeonFloors().forEach { floor ->
            floor.allRooms.forEach { room ->
                room.getFeatureList().forEach { feature ->
                    if (feature is StairsFeature) {
                        stairsToDecide.add(feature)
                    }
                }
            }
        }

        stairsToDecide.forEach { it.getConnectedRoom() }

        getDungeonFloors().forEach { it.complete() }
    }

    fun getRoomAt(floor: Int, x: Int, y: Int): Room? {
        getDungeonFloorAtLevel(floor)?.allRooms?.forEach {room ->
            if (room.containsGlobalPoint(Point(x, y))) return room
        }
        return null
    }

    fun getDungeonFloors(): List<Floor> {
        return dungeonFloors
    }

    fun getGlobalModifier() = globalModifier

    fun getMemoryItem(): DungeonHistoryItem {
        return DungeonHistoryItem(
                name,
                seed,
                id,
                id
        )
    }
}